### DISCUSSION TITLE
Referências de estudos de caso de uso avançado do GitHub Actions (ilimitado e gratuito para Open Source) inexplorado no Brasil; foco cross-platform

### DISCUSSION THREAD

**Discussion by fititnt** [frontendbr/forum#2182] [Mostre seu trabalho]
Salve salve, comunidade brasileira que cria tendências! Antes mesmo do _GitHub discussions_ já era a maior do seu tipo. Coincidência? Acho que não xD.

Nós do @HXL-CPLP estamos testando/documentando com <3 como é possível usar GitHub Actions (que rodam em máquinas poderosas) para **automatização completa** até mesmo para publicar nos marketplaces ou (tanto para testes como para quem não tem $ taxas) pelo menos gerar _night builds_ baixáveis na aba releases do próprio GitHub para todas as plataformas que tecnologias típicas conseguiram ser empacotadas, mesmo que você como desenvolvedor não tenha todas as combinações de sistemas operacionais e SDKs de mobile.

No momento três formas de empacotar estão sendo documentadas (note que algumas permitem mais linguagens e frameworks):

- dotnet (inspirado por @saraiva1989)
- Electron (inspirado por @betafcc)
- Flutter (inspirado por @rodriguesJeff)
- _Sugestões?_

#### Explicação curta

Na imagem, uma busca por GitHub actions aqui no frontendbr/forum retorna isso (resumindo: discussões principalmente da época que era beta; hoje a implementação pode ser usada em produção, mas pessoas não estão atentas a isso).

![Captura-de-tela-de-2021-10-27-19-24-34](https://user-images.githubusercontent.com/812299/139156283-d06be8b0-b24f-44d8-bb85-c6b4f647621d.png)

Numa olhada em grupos de Facebook, pelo menos em 2021-10-27, a maioria de algo mais sério acontece naqueles relacionados a DevOps. Referências onde tem potencial de ajudar (como quem desenvolve software) também é rara.

Coloquei a imagem aqui porque é necessário dar um pontapé no assunto.

#### Explicação longa

O que digo de _potencial inexplorado_ é que os [executores hospedados no GitHub](https://docs.github.com/pt/actions/using-github-hosted-runners/about-github-hosted-runners), além de ter imagens pre-prontas com dependencias (o que permite rodar em < 60s coisas que em outros C.I. poderia demorar minutos antes de começar) tem 7GB de ram (Linux, Windows) e 14 GB de RAM (macOS, que inclusive tem aceleração gráfica habilitada). Enquanto o soft-limit 50mb e hard limit (100GB) de arquivos em repositório são conhecidos, o hard limit de [binários nos releases do GitHub, como os que que podem ser gerados automaticamente é de (em 2021-10-27) 2 GB](https://docs.github.com/en/repositories/releasing-projects-on-github/about-releases). Embora seja possível deletar pacotes para ter mais espaço em discos (o que é garantido não menos do que 14GB livres, por isso muita coisa é empacotada como outras actions), é possível compilar aplicativos de Android e iOS, mas não tem disco suficiente para compilar algo como sistema operacional Android inteiro (recomendado 250GB ou mais). Também existe um "cache temporário" opcional, otimizado para dados mais acessados entre builds por repositório de 5GB, que é extremamente.

> Sobre o que podem perguntar: **qual a pegadinha? Porque é de graça? Por que liberam?** Os termos de uso podem contactar amigavelmente repositórios que estiverem fazendo algo incomum, mas, na prática, principal preocupação com abusos é... contas novas usando GitHub (e, principalmente infraestrutura da Azure) para minerar criptomoedas coisa que normalmente ocorre em repositórios privados OU gente que dava pull request em projetos dos outros e injetava minerador (agora tem opção para exigir humano aprovar rodar testes).

Ou seja: existe margem para, colaborativamente, criar esse tipo de documentação de como ajudar quem faz open source a ter maior controle com testes **inclusive de apps gráficos em outros sistemas operacionais que ela não possuí**! Comparado a quem fazia anos atrás usando Travis CI ou Azure Pipelines, ainda que uma e outra funcionalidade não tenha equivalente, não só ficou muito mais fácil (requer saber YAML, preferencialmente com guias de ajuda, que vocês poderiam cria em cima dos testes que estamos fazendo no @HXL-CPLP). Daí que uma ideia desse tópico é trocar ajudas.

Fato não intuitivo: a forma como Windows Server e macOS estão otimizados (por já ter pacotes instalados) rodar testes e gerar binários pode ser mais rápido do que fluxos de trabalho usando linguagens scripts que não fazem cache ou que usam matriz de testes exageradas. As GitHub Actions reusáveis para dotnet (e tudo relacionado a Microsoft) e Xcode (tanto app desktop macOS, como para compilar para iPhone) tão super otimizadas e documentadas, em alguns casos até melhor do que as de Linux, visto que as empresas pagam funcionários para manter elas perfeitas.

Para finalizar este resumo rápido: a ideia toda de ter GitHub Actions gratuita (e empresas quererem facilitar para desenvolvedores não precisarem máquinas poderosas locais para fazer Software no sistema operacional delas) faz muito sentido com iniciativas como editar código usando navegador (vide https://github.dev e https://vscode.dev). **E fomentar a ideia de troca de ajudas de como usar actions (de forma eficiente) é útil para comunidade em geral (mas especialmente para quem não tem todo tipo de hardware)**. Some a isso que ao fazer prontas de conceito com dotnet + Electron + Flutter tem coisa que dá para reusar do ponto de vista de aprendizado (mesmo que cada um usa o que preferir!)

### Exemplos de feedback incentivados neste post

Eu sei que post ficou longo, então vou propor já de início pelo menos essas sugestões. Faço isso

- **Você direta ou indiretamente participa de projeto open source que já usa GitHub Actions?**
    - Se sim, fale a respeito!
    - Se você tem projeto em GitHub Actions e alguma outra alternativa (que não seja paga) pode também falar sobre impressões pessoais
- **(Foco especial) quem já usa GitHub actions para criar builds cross-platform de apps com interface gráfica: apresente-se!**
    - Inclusive desktop: Windows, Linux (diversos marketplaces), e macOS
    - Inclusive móbile: web (diversas formas de empacotar), Android, iOS (com ou sem envio para Apple Store)
- **Para quem tem projeto open source (mesmo que pequeno, mas que tenha tido atividade nos últimos 12 meses), que poderia ter automação para gerar builds (mas é processo manual)**
    - Caso tenha interesse, pode comentar aqui. A gente do @HXL-CPLP pode te ajudar com isso. Motivo: como existem muitas plataformas diferentes, vale a pena ajudar pelo menos de 1 a 3 projetos reais.
        - Nota: projetos que lidem com credenciais externas de acesso requerem certos cuidados. Quem quiser ajuda vale a pena ter contato por privado para tirar dúvidas. Existem coisas que apenas quem é dono de um repositório pode fazer.
- **Alguém conhece se brasileiro já publicou alguma GitHub Actions**
    - dica para quem gostar da idéia: as que são feitas em TypeScript (diferente das em Docker, que requer hospedeiro em Linux) por padrão podem rodar em Windows, macOS e Linux. Logo TypeScript é especialmente útil se surgirem necessidades mais específicas!

-------

Nota: possivelmente esse tópico inicial seja editado caso sugestões novas surjam.

**Comment by saraiva1989** [comment]
Eu uso github actions, mas na verdade não foi eu quem criou.
Quando subo uma aplicação para o azure e o código está hospedado no github, eu consigo vincular meu repositorio com o azure. Sendo assim o prório azure gera um arquivo .yml no qual tem os parametros para o deploy.

Eu não faço ideia como montar os scripts de deploy de forma manual, apenas conheço integrando direto o github com azure.

**Reply by fititnt** [reply]
> Editado: **agora** eu vi o que você queria dizer:
> - Em https://github.com/saraiva1989/gamereleasevue/commit/11930dbc2c33e2d2d4b30ac49cd30c416f990852 tem um exemplo de commit de @msftgits que um ano atrás criou o actions para você. Interessante isso.
>     - Numa olhada rápida, aquele action do ano passado, usa `uses: Azure/static-web-apps-deploy@v0.0.1-preview`; Por esse nome, eu não duvidaria que esse action nem esteja mais funcionando hoje em dia e teria que ser manualmente resolvido.

---

Eu sei que você não criou xD a action (a do [dotnet é até oficial](https://github.com/marketplace/actions/setup-net-core-sdk)). Por isso falei inspiração (do ponto de vista de necessidade, visto que é uma tecnologia que tem gente usando). Dias depois os outros dois colegas sugeriram outras abordagens (exceto o Rodrigues, que também falou (acho) C\#).

Meu ponto de vista aqui do post, embora sempre vai ter gente que quer uma estratégia mais específica, mas:

1. Actions do GitHub, que já são production ready, são pouco conhecidas.
    - **Na verdade, provavelmente quem veio do .NET deve estar bem mais acostumado com GitHub actions do que os demais**: a Microsoft fez um bom trabalho.
    - O flutter é muito novo, mas não sei se tem muita documentação de como podia ser feito teste completo de apps Electron com alternativas antes do GitHub action, como o Travis CI.
2. Se remover as partes mais específicas de cada plataforma (ou seja, o que é extremamente especifico de dotnet vs flutter vs electron vs (...)) o fluxo de trabalho de como usar actions para por projeto em produção acaba tendo conceitos reusáveis.
3. Mesmo actions mais específicas de cada plataforma, não raro até a organização dos parâmetros delas são parecidas.
    - Então, implicitamente, a abstração provida por usar YAML para abstrair deploy e testagem permite discussão "mais alto nível" do tipo ter blocos de montar de como pessoal pode ter boas práticas em fazer deploy/testagem
    - Coisas que não são reusáveis (mas fica meio claro): algumas abordagens vão usar um `- run` diferente:
        - **dotnet**: `- run: dotnet test programma-exemplum-dotnet.csproj`
        - **electron**: `run: npm test` (nota: isso varia muito; menos padrão que outros)
        - **flutter**: `- run: flutter test`
  - Ou seja, mesmo nas coisas que não são iguais, da para ter uma ideia do que está acontecendo.

**Comment by wilsonneto-swe** [comment]
Aproveitando o tópico também, esta semana fiz uma live super massa com o pessoal do Azure na Prática onde a gente subiu aplicação Next.js tanto em um storage estático com Github Actions quanto também automatizamos a publicação da mesma com containers para quando for necessário servidor:

Aula Next.js + GH Actions + Azure - Automatizando deploy do Next.js:
https://www.youtube.com/watch?v=Vuu2IwTyOiY

Quem se interessar também, segue lá na twitch para trocarmos ideas, umas 3 noites por semana estarei ali ao vivo trocando ideias:
https://www.twitch.tv/wilsonnetodev

E amanhã às 21h farei uma live com amigos sobre as novidades do Next.js.
